Allow multiple submissions in modeling exercises

maxr96 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We want to allow students to submit exercise until the exercise deadline.
The are following criteria for allowing to submit an exercise are the following:
    - There is no due date
    - The due date is before the current date
    - The participation initialization date is after the due date
    - The result is not yet set (avoid infinite submission for late and without due date submissions)

### Description
<!-- Describe your changes in detail -->
The Save button is gone, but the student can submit as much as needed and last submission before the exercise due date will be reviewed. If student started participating in the exercise after the due date the submit button will be called `Submit(deadline missed)`. Tooltips added for both buttons. 

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Overview
3. Select Course with Modelling Exercise
4. Start new Modeling Exercise
5. Submit exercise multiple time
6. Wait for exercise deadline and check that Submit button is disabled now

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![Peek 2019-11-09 15-27](https://user-images.githubusercontent.com/17262376/68530121-9a7fad80-0305-11ea-99fc-a61dc999e1fc.gif)

krusche commented:
Auto-save should still invoke the save REST call and NOT the submit call.
Please double check the transition to participation.initializationState = Finished.

krusche commented on src/main/webapp/app/modeling-submission/modeling-submission.component.ts:
```diff
@@ -224,106 +229,73 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
+                () => (this.isSaving = false),
+            );
+        } else {
+            this.modelingSubmissionService.create(this.submission, this.modelingExercise.id).subscribe(
```
why did you delete update here?
I think we should always use update instead of create

maxr96 replied:
I didn't delete it. I only check if we already have submission, then we update it, if not then create.

krusche commented on src/main/webapp/i18n/de/global.json:
```diff
@@ -198,7 +198,10 @@
             "scores": "Bewertungen",
-            "manageHints": "Verwalte Hinweise"
+            "manageHints": "Verwalte Hinweise",
+            "submitTooltip": "Du kannst mehrmals abgeben. Deine Bewertung ergibt sich immer aus der letzten Abgabe.",
```
include "vor der Abgabefrist"
also use "Abgabefrist" instead of "Deadline"

krusche commented on src/main/webapp/i18n/en/global.json:
```diff
@@ -204,7 +205,10 @@
             "scores": "Scores",
-            "manageHints": "Manage Hints"
+            "manageHints": "Manage Hints",
+            "submitTooltip": "You can submit multiple times. Your latest submission will count towards your score.",
```
"before the deadline"

krusche commented on src/test/java/de/tum/in/www1/artemis/ModelingAssessmentIntegrationTest.java:
```diff
@@ -376,20 +376,6 @@ public void automaticAssessmentUponModelSubmission_noResultInDatabase() throws E
-    @Test
-    @WithMockUser(username = "student2")
-    public void noAutomaticAssessmentUponModelSave() throws Exception {
-        saveModelingSubmissionAndAssessment(true);
```
why do you delete this test case? It is still valid!

maxr96 commented on src/main/webapp/app/modeling-submission/modeling-submission.component.ts:
```diff
@@ -494,4 +466,11 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
+     * The exercise is still active if it's due date hasn't passed yet, or if the participation was started after the due date has already passed
+     */
+    get isActive() {
+        return this.modelingExercise && (!this.modelingExercise.dueDate || moment(this.modelingExercise.dueDate).isAfter(moment()));
```
Yes but I do it with isLate property. So if the exercise was started after the due date, student will see `late submission` button

ungaralex replied:
From our Slack discussion: 
> except the participation start date is after the exercise due date, then they can submit forever

Also, please change the comment then, because there you say that `isActive` checks if the participation was started after the due date. It's fine if you use the `isLate` property, but then the JSDoc should reflect that.

maxr96 replied:
@ungaralex True, I forgot to update JSDoc after the latest changes.

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:
```diff
@@ -204,15 +204,9 @@ public ModelingSubmission save(ModelingSubmission modelingSubmission, ModelingEx
-                    "participationExists");
+        final var exerciseDueDate = modelingExercise.getDueDate();
+        if (exerciseDueDate != null && exerciseDueDate.isBefore(ZonedDateTime.now()) && participation.getInitializationDate().isBefore(exerciseDueDate)) {
+            throw new BadRequestAlertException("You missed the deadline for submitting modeling exercise", "modelingSubmission", "idexists");
```
Sure

maxr96 commented:
> Good job implementing this feature @maxr96
> I have tested it and the functionality and requests that are sent looked good to me.
> 
> However there is still an error that is thrown when you initially open the apollon-editor
> ![image](https://user-images.githubusercontent.com/33764166/67437213-9ad33580-f5f0-11e9-99e7-b4ada29a11f5.png)

Thanks for checking @lovaiible. Seems like change detection where triggered multiple times before we actually get the modeling exercise. I added an explicit check to avoid receiving the error

maxr96 commented:
> Thanks for the help. Looks good to me! You should only have a look at you `isActive` property again

Thanks for checking. I looked again at `isActive`, but for me it covered all the cases. I added a note however about the result that we don't check

maxr96 commented:
> Auto-save should still invoke the save REST call and NOT the submit call.
> Please double check the transition to participation.initializationState = Finished.

@krusche I see, I returned the saving functionality and auto-save calls saving endpoint again. Also since the saving is relevant again I added back the removed integration test. 
Regarding the `participation.initializationState = Finished` transition I didn't notice any strange behavior. After the student click submit he will only see `Submitted, waiting for grading.` message on page reload.

maxr96 commented on src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:
```diff
@@ -204,15 +205,9 @@ public ModelingSubmission save(ModelingSubmission modelingSubmission, ModelingEx
-            throw new BadRequestAlertException("User " + username + " already participated in exercise with id " + modelingExercise.getId(), "modelingSubmission",
-                    "participationExists");
+        final var exerciseDueDate = modelingExercise.getDueDate();
+        if (exerciseDueDate != null && exerciseDueDate.isBefore(ZonedDateTime.now()) && participation.getInitializationDate().isBefore(exerciseDueDate)) {
```
You are totally right here! Didn't think about this

maxr96 commented on src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:
```diff
@@ -401,6 +409,39 @@ public void getModelSubmissionForModelingEditor() throws Exception {
+
+        final var submissionInDb = modelingSubmissionRepo.findById(submittedSubmission.getId());
+        assertThat(submissionInDb.isPresent());
+        assertThat(submissionInDb.get().getModel()).isEqualTo(emptyModel);
```
True!

maxr96 commented:
> I reviewed your changes and it looks good to me! I have added some minor comments, can you have a look at them?

Thank you for checking @lovaiible . I implemented the suggested changes

maxr96 commented:
>     * It still says, I have unsaved changes (alert) even after submitting and trying to leave the page.
> 
>     * It also says late submission in the upper right corner above the submit button, although the deadline hasn't expired, yet.
> 
> 
> Apart from that, the rest looks fine +1

Thank you for checking my PR.
1. You are right. Had to debug it a bit to find out that the size was saved differently when reassigning the model on updating the submission. Not it should work as expected.
2. I first checked and it was only saying late submission if the assessment due date is passed or it is null. But after submitting I also had this strange behavior and I figured out that we are missing submissions on the participation. I reconnected them and now the string is the same as on course-exercise page.

krusche commented:
Can you please update the description and add a screenshot / screencast?

maxr96 commented:
@krusche I updated the description. Added a screen cast showing multiple submission in action. Also added multiple client tests to check if the submission is allowed or blocked properly

krusche commented:
@maxr96 can you please resolve the merge conflicts?
